Adding profiling capability to c++ ddp collective functions

mrzzd opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46471 Adding profiling capability to c++ ddp collective functions**

Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

Commit:
Adding profiling capability to c++ ddp collective functions



Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

pritamdamania87 commented:
Can you update the PR summary to include an example of what the profiling output looks like with your changes?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1229,6 +1225,25 @@ def test_all_reduce_result_cuda(self):
+                work = op(*args, async_op=async_op, **kwargs)
+                if async_op:
+                    work.wait()
+                    work._get_profiling_future().wait()
```
It looks like `_get_profiling_future` is only used here? If so, why do we need this? Once we exit the the profiler context, the profiling information should be recorded appropriately.

rohan-varma replied:
I don't think that's necessarily the case, as `work.wait()` could return without the profiling callback having ran and this wait ensures that the profiling callback (one that terminates the record function)is finished successfully. We had to do something similar with RPC (see: https://github.com/pytorch/pytorch/pull/38352/files) although in that case it is transparent to the user.

Ideally when there's profiling, similar to RPC, work.wait() should ensure the profiling callbacks have ran before returning. It might depend on future/work merge, though maybe we can get it to work now by modifying `::wait()` to await the profiling future if one exists

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -119,10 +135,19 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroup::Work::getFuture() {
   std::unique_lock<std::mutex> lock(mutex_);
   completed_ = true;
   exception_ = exception;
+  if (recordFunctionEndCallback_) {
+    recordFunctionEndCallback_();
```
Shouldn't we do this for `finishAndThrow` below as well?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1036,19 +1036,15 @@ def _test_reduce_helper(
+                tensor = _build_tensor(src + 1).fill_(master_value if rank == src else worker_value)
+                if cuda:
+                    tensor = tensor.cuda(rank_to_GPU[rank][0])
+                self.call_dist_op("reduce", async_op, dist.reduce, tensor, src, op, group_id)
```
We should also test that `send` and `recv` work as well for both gloo and NCCL.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1038,13 +1041,23 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   work->outputs_ = std::make_shared<std::vector<at::Tensor>>(outputs);
   work->futureNCCLCallbackStreams_ = futureNCCLCallbackStreams_;
 
+  if (work->recordFunctionEndCallback_) {
```
We probably need to enhance `pointTopoint` as well to cover send and recv?

rohan-varma replied:
I think it's fine to add as a follow up PR, but can we file a GH issue for this (and any other follow up tasks?)

rohan-varma replied:
https://github.com/pytorch/pytorch/issues/47482

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -172,7 +173,14 @@ class AllreduceNCCLTest : public NCCLTest {
     valueInitialization();
 
-    return pg_->allreduce(tensors_);
+    using namespace torch::autograd::profiler;
+    // Make sure enabling profile does not make any issue. Note, in single
+    // process multi-device mode we do not expect any events be populated for
+    // collective operations.
+    enableProfiler({ProfilerState::CPU});
+    auto results = pg_->allreduce(tensors_);
+    disableProfiler();
+    return results;
```
Do we need to add a cpp test since we're already covered by python tests?

rohan-varma replied:
I think it would be useful from profiler perspective to just test the C++ API as well, which skips the parsing/event aggregation logic in profiler that happens in python. We have similar tests in https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_misc.cpp#L2185-L2198

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -9,6 +9,7 @@
 #include <c10/cuda/CUDAGuard.h>
```
We probably need to add tests to ProcessGroupMPITest to validate the profiling works correctly for that as well.

pritamdamania87 reviewed (commented):
Thanks for working on this, the overall structure looks good!

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1022,7 +1022,17 @@ that adds a prefix to each key inserted to the store.
+          py::call_guard<py::gil_scoped_acquire>(),
+          R"(
+        Returns future that completes when the profiling event corresponding
+        to the creation of this RRef on the remote node has been recorded.
```
This isn't the right doc for this function right? There's no RRef here.

rohan-varma commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -51,10 +53,24 @@ bool isP2POp(OpType opType) {
+    auto recordingFunction = std::make_shared<at::RecordFunction>(at::RecordScope::USER_SCOPE);
+    if (recordingFunction->active) {
+        recordingFunction->before(profiling_title, {});
+        std::function<void()> end_handler = [this, recordingFunction]() {
```
Can we `std::move(recordingFunction)` since it's not used after this anymore to avoid a copy?

mrzzd replied:
It is shared_ptr and copy is cheap. right after this block also the extra copy is destroyed. If I wanted to use std:move line 62 will change to something much less readable:
` std::function<void()> end_handler = [this, recordingFunction{std::move(recordingFunction)}]()`

rohan-varma commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -156,6 +158,9 @@ class ProcessGroup {
     // Operation type that this work object refers to.
     OpType opType_;
+
+    std::function<void()> recordFunctionEndCallback_;
```
Nit: can we add some comments about the purposes of these 2 variables.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1038,13 +1041,23 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+    // recordFunctionEndCallback_ is normally called in fininsh() function by
+    // base class, but since finish is not called by WorkNCCL, we schedule this
+    // function to be run when work is done.
+    work->getFuture()->addCallback(std::move(work->recordFunctionEndCallback_));
```
I guess we would just need to change the if statement from "`f (work->recordFunctionEndCallback_) {` to `if (work->recordFunctionEndCallback_ && can_profile) {`

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1476,7 +1497,8 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
         },
-        OpType::ALLTOALL_BASE);
+        OpType::ALLTOALL_BASE,
+        "all_to_all");
```
Just to confirm, this won't include shape information for now right? That's fine for this diff but just wanted to make sure

rohan-varma commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -172,7 +173,14 @@ class AllreduceNCCLTest : public NCCLTest {
+    enableProfiler({ProfilerState::CPU});
+    auto results = pg_->allreduce(tensors_);
+    disableProfiler();
+    return results;
```
Generally a test should assert something/some condition. Could we search through `results` and verify there is an allreduce here? You can see https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_misc.cpp#L2185-L2198 as an example.

rohan-varma replied:
Can we do a single device per process test somehow? In the current version are the profiling results empty? 

Also tangentially related - can you add a comment somewhere appropriate that specifies that it only works for single process per device?

rohan-varma replied:
Following up here, is it possible to add some asserts on the expected result?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1229,6 +1225,25 @@ def test_all_reduce_result_cuda(self):
+                events = [event for event in prof.function_events if partial_key in event.name]
+                return events[0] if len(events) > 0 else None
+
+            recv_event = get_event(profiling_title)
```
Nit: here we've matched on a partial key, can we also add an assert for what the exact name would look like?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1229,6 +1225,25 @@ def test_all_reduce_result_cuda(self):
+                events = [event for event in prof.function_events if partial_key in event.name]
+                return events[0] if len(events) > 0 else None
+
+            recv_event = get_event(profiling_title)
```
Nit: Is it always a `recv` event or are there different type of collective comm. calls here? If the latter is true can we have a name such as `comm_event` which might be less confusing?

rohan-varma replied:
Yes, that is fine. I was mostly asking because I was curious why it was named `recv_event`.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1229,6 +1225,25 @@ def test_all_reduce_result_cuda(self):
+
+            recv_event = get_event(profiling_title)
+            if expect_event:
+                self.assertEqual(recv_event.count, 1)
```
Can we have a test where we do > 1 collective comm of the same type, and then > 1 collective comm of different types and validate the counts for those as well?

rohan-varma reviewed (commented):
Awesome, this is looking great overall! Left some comments inline. Could you also paste what the profiling output looks like in the PR description (you can get that with `print(prof.key_averages().table())` in one of the tests)?

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -131,6 +156,10 @@ void ProcessGroup::Work::finishAndThrow(std::exception_ptr exception) {
   exception_ = exception;
+  if (recordFunctionEndCallback_) {
+    recordFunctionEndCallback_();
+    recordFunctionEndCallback_ = nullptr;
```
Do we have tests that exercise this code path (`finishAndThrow`)?

rohan-varma replied:
@mrzzd Just following up here - do we need to add these tests?

rohan-varma commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -137,6 +135,10 @@ class ProcessGroup {
 
     OpType retrieveOpType();
 
+    // Keeps track of the future responsible for profiling owner creation
```
what does "profiling owner creation" mean? Do you just mean that this is a future that is complete when the profiling has finished?

rohan-varma commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -156,6 +158,11 @@ class ProcessGroup {
     // Operation type that this work object refers to.
     OpType opType_;
+
+    // When profiling, the call back to record end of operation event. This
```
nit: "callback" instead of "call back"

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1038,13 +1041,25 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+  }
+
+
+
```
nit: unneeded extra lines?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1176,6 +1171,64 @@ def test_reduce_full_group_max(self):
             self._test_reduce_helper(group, group_id, rank, dist.ReduceOp.MAX, -1, 10, 10)
 
+        # REDUCE TWICE
+        def _rest_reduce_twice_helper(
```
Nit: What does `_rest_reduce_twice` mean? Should it be `_test_reduce_twice`?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1223,6 +1276,31 @@ def test_all_reduce_result_cuda(self):
+                if is_async:
+                    for work in works:
+                        work.wait()
+                        work._get_profiling_future().wait()
```
I'm assuming that the test is flaky in the case where we remove this call? Is it okay to ship this prototype because we would need this explicit wait call from user code?

cc @pritamdamania87 - I guess we might be able to hack something but in the long term this will probably depend on future/work merge and we would implement this by adding a then callback like we do for RPC. Do you have any thoughts on what we can do currently?

rohan-varma reviewed (commented):
Thank you for the awesome work! It looks great, but some larger things I think we should talk about: 
1) ProcessGroupMPI tests (looks like nccl and gloo are thoroughly tested)
2) send/recv profiling follow up, and file GH issues for that and any other follow up tasks
3) Discuss a design for removing `work._get_profiling_future().wait()` before we expose this to users. 
4) Could you also add the profiling output to the PR summary?

rohan-varma commented:
Discussed about (3) offline, we will try to remove the call and the profiling should still be done transparently since the nccl callback should be invoked inline.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1282,6 +1373,27 @@ def test_all_reduce_sum_cuda(self):
 
+        @unittest.skipIf(
+            BACKEND != "gloo" and BACKEND != "nccl",
+            "Only Gloo backend will have CUDA allReduce tested",
```
The test runs if backend is NCCL or Gloo, so this message needs to be updated right?

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM overall, thanks for doing this! Had a couple of nits and 2 comments about testing. Feel free to land after taking a look at those.

Commit:
Update on "Adding profiling capability to c++ ddp collective functions"


Differential Revision: [D23948397](https://our.internmc.facebook.com/intern/diff/D23948397/)

[ghstack-poisoned]
